### PR TITLE
[SPARK-16104][SQL] Do not creaate CSV writer object for every flush when writing

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR let `CsvWriter` object is not created for each time but able to be reused. This way was taken after from JSON data source.

Original `CsvWriter` was being created for each row but it was enhanced in https://github.com/apache/spark/pull/13229. However, it still creates `CsvWriter` object for each `flush()` in `LineCsvWriter`. It seems it does not have to close the object and re-create this for every flush. 

It follows the original logic as it is but `CsvWriter` is reused by reseting `CharArrayWriter`. 
## How was this patch tested?

Existing tests should cover this.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 1]
```diff
@@ -177,38 +180,61 @@ private[sql] class CsvOutputWriter(
...
+    format.setComment(options.comment)
+
+    writerSettings.setNullValue(options.nullValue)
+    writerSettings.setEmptyValue(options.nullValue)
+    writerSettings.setSkipEmptyLines(true)
+    writerSettings.setQuoteAllFields(false)
+    writerSettings.setHeaders(headers: _*)
+    writerSettings.setQuoteEscapingEnabled(options.escapeQuotes)
+
+    new CsvWriter(writer, writerSettings)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Please let me organise this in a separate class at the PR for refactoring this.

**Comment by HyukjinKwon** [comment]
Hm.. I will try to minimise the changes and move all the classes in another PR.

**Comment by HyukjinKwon** [comment]
Hi @davies, it seems the past change in your PR is related with this. Could you please take a look?

**Comment by HyukjinKwon** [comment]
Hi @davies, it seems the past change in your PR is related with this. Could you please take a look?

**Comment by davies** [comment]
LGTM, 
Merging this into master, thanks!

**Pull request closed without merging**